service/history/queues/slice.go

```diff
@@ -0,0 +1,370 @@
+	scope Scope,
+) *SliceImpl {
+	return &SliceImpl{
+		executableInitializer:  executableInitializer,
+		scope:                  scope,
+		outstandingExecutables: make(map[tasks.Key]Executable),
+		iterators: []Iterator{
+			NewIterator(paginationFnProvider, scope.Range),
+		},
+		destroyed: false,
+	}
+}
+
+func (s *SliceImpl) Scope() Scope {
+	s.validateNotDestroyed()
+	return s.scope
+}
+
+func (s *SliceImpl) CanSplitByRange(key tasks.Key) bool {
+	s.validateNotDestroyed()
+	return s.scope.CanSplitByRange(key)
+}
+
+func (s *SliceImpl) SplitByRange(key tasks.Key) (leftSlice Slice, rightSlice Slice) {
+	if !s.CanSplitByRange(key) {
+		panic(fmt.Sprintf("Unable to split queue slice with range %v at %v", s.scope.Range, key))
+	}
+
+	leftScope, rightScope := s.scope.SplitByRange(key)
+	leftExecutables, rightExecutables := s.splitExecutables(leftScope, rightScope)
+
+	leftIterators := make([]Iterator, 0, len(s.iterators)/2)
+	rightIterators := make([]Iterator, 0, len(s.iterators)/2)
+	for _, iter := range s.iterators {
+		iterRange := iter.Range()
+		if leftScope.Range.ContainsRange(iterRange) {
+			leftIterators = append(leftIterators, iter)
+			continue
+		}
+
+		if rightScope.Range.ContainsRange(iterRange) {
+			rightIterators = append(rightIterators, iter)
+			continue
+		}
+
+		leftIter, rightIter := iter.Split(key)
+		leftIterators = append(leftIterators, leftIter)
+		rightIterators = append(rightIterators, rightIter)
+	}
+
+	leftSlice = &SliceImpl{
+		executableInitializer:  s.executableInitializer,
+		scope:                  leftScope,
+		outstandingExecutables: leftExecutables,
+		iterators:              leftIterators,
+	}
+	rightSlice = &SliceImpl{
+		executableInitializer:  s.executableInitializer,
+		scope:                  rightScope,
+		outstandingExecutables: rightExecutables,
+		iterators:              rightIterators,
+	}
+
+	s.destroy()
+	return leftSlice, rightSlice
+}
+
+func (s *SliceImpl) SplitByPredicate(predicate tasks.Predicate) (passSlice Slice, failSlice Slice) {
+	s.validateNotDestroyed()
+
+	passScope, failScope := s.scope.SplitByPredicate(predicate)
+	passExecutables, failExecutables := s.splitExecutables(passScope, failScope)
+
+	passIterators := make([]Iterator, 0, len(s.iterators))
+	failIterators := make([]Iterator, 0, len(s.iterators))
+	for _, iter := range s.iterators {
+		passIterators = append(passIterators, iter)
+		failIterators = append(failIterators, iter.Remaining())
+	}
+
+	passSlice = &SliceImpl{
+		executableInitializer:  s.executableInitializer,
+		scope:                  passScope,
+		outstandingExecutables: passExecutables,
+		iterators:              passIterators,
+	}
+	failSlice = &SliceImpl{
+		executableInitializer:  s.executableInitializer,
+		scope:                  failScope,
+		outstandingExecutables: failExecutables,
+		iterators:              failIterators,
+	}
+
+	s.destroy()
+	return passSlice, failSlice
+}
+
+func (s *SliceImpl) splitExecutables(
+	thisScope Scope,
+	thatScope Scope,
```

Comment:
nit: I like left & right better than this & that.

Reply:
It's not always left & right since we can split by predicate. so used this, that here.